db/python/layers/cohort.py

```diff
@@ -0,0 +1,137 @@
+import json
+
+from db.python.connect import Connection
+from db.python.layers.base import BaseLayer
+from db.python.layers.sequencing_group import SequencingGroupLayer
+from db.python.tables.analysis import AnalysisTable
+from db.python.tables.cohort import CohortFilter, CohortTable, CohortTemplateFilter
+from db.python.tables.project import ProjectId, ProjectPermissionsTable
+from db.python.tables.sample import SampleTable
+from db.python.tables.sequencing_group import (
+    SequencingGroupFilter,
+    SequencingGroupTable,
+)
+from db.python.utils import GenericFilter, get_logger
+from models.models.cohort import Cohort, CohortCriteria, CohortTemplate
+from models.utils.sequencing_group_id_format import (
+    sequencing_group_id_transform_to_raw_list,
+)
+
+logger = get_logger()
+
+
+class CohortLayer(BaseLayer):
+    """Layer for cohort logic"""
+
+    def __init__(self, connection: Connection):
+        super().__init__(connection)
+
+        self.sampt = SampleTable(connection)
+        self.at = AnalysisTable(connection)
+        self.ct = CohortTable(connection)
+        self.pt = ProjectPermissionsTable(connection)
+        self.sgt = SequencingGroupTable(connection)
+        self.sglayer = SequencingGroupLayer(self.connection)
+
+    async def query(self, filter_: CohortFilter) -> list[Cohort]:
+        """Query Cohorts"""
+        cohorts = await self.ct.query(filter_)
+        return cohorts
+
+    async def query_cohort_templates(self, filter_: CohortTemplateFilter) -> list[CohortTemplate]:
+        """Query CohortTemplates"""
+        cohort_templates = await self.ct.query_cohort_templates(filter_)
+        return cohort_templates
+
+    async def get_cohort_sequencing_group_ids(self, cohort_id: int) -> list[int]:
+        """
+        Get the sequencing group IDs for the given cohort.
+        """
+        return await self.ct.get_cohort_sequencing_group_ids(cohort_id)
+
+    async def create_cohort_template(
+            self,
+            cohort_template: CohortTemplate,
+            project: ProjectId,
+
+    ):
+        """
+        Create new cohort template
+        """
+
+        return await self.ct.create_cohort_template(
+            name=cohort_template.name,
+            description=cohort_template.description,
+            criteria=dict(cohort_template.criteria),
+            project=project
+        )
+
+    async def create_cohort_from_criteria(
+            self,
+            project_to_write: ProjectId,
+            author: str,
+            description: str,
+            cohort_name: str,
+            cohort_criteria: CohortCriteria = None,
+            template_id: int = None,
+    ):
+        """
+        Create a new cohort from the given parameters. Returns the newly created cohort_id.
```

Comment:
Pretty obvious, but documentation should probably say “One of `cohort_criteria` or `template_id` must be provided” or so.

Reply:
Interestingly I was thinking about whether we add functionality to add both. I.e., if you specify a template + new criteria it will combine together (and generate a new template)

Reply:
Yes, I wondered if that could usefully mean something. The case I ran into was setting both to None makes it explode 💥

Reply:
Yes sorry, let me push some code to properly handle the different cases!!

Reply:
Not complaining, exploding was a reasonable response to the nonsensical input :smile:

Reply:
Haha perhaps it should fail more gracefully though :P